File: solutions/C++/coin_change_dp_array.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

using namespace std;

/**
 * Find minimum number of coins needed to make amount using DP array.
 *
 * Time Complexity: O(n * amount)
 * Space Complexity: O(amount)
 */
int coinChangeDP(vector<int>& coins, int amount) {
    vector<int> dp(amount + 1, INT_MAX);
    dp[0] = 0;

    for (int i = 1; i <= amount; i++) {
        for (int coin : coins) {
            if (coin <= i && dp[i - coin] != INT_MAX) {
                dp[i] = min(dp[i], dp[i - coin] + 1);
            }
        }
    }

    return dp[amount] == INT_MAX ? -1 : dp[amount];
}

int main() {
    vector<int> coins1 = {1, 2, 5};
    cout << coinChangeDP(coins1, 5) << endl;  // 1

    vector<int> coins2 = {2};
    cout << coinChangeDP(coins2, 3) << endl;  // -1

    return 0;
}

File: solutions/Golang/coin_change_dp_array.go
package main

import (
	"fmt"
	"math"
)

/*
 * Find minimum number of coins needed to make amount using DP array.
 *
 * Time Complexity: O(n * amount)
 * Space Complexity: O(amount)
 */
func coinChangeDP(coins []int, amount int) int {
	dp := make([]int, amount+1)
	for i := 1; i <= amount; i++ {
		dp[i] = math.MaxInt32
	}
	dp[0] = 0

	for i := 1; i <= amount; i++ {
		for _, coin := range coins {
			if coin <= i && dp[i-coin] != math.MaxInt32 {
				if dp[i-coin]+1 < dp[i] {
					dp[i] = dp[i-coin] + 1
				}
			}
		}
	}

	if dp[amount] == math.MaxInt32 {
		return -1
	}
	return dp[amount]
}

func main() {
	fmt.Println(coinChangeDP([]int{1, 2, 5}, 5))   // 1
	fmt.Println(coinChangeDP([]int{2}, 3))        // -1
}
